contracts/token/ERC20/ERC20Channel.sol

```diff
@@ -0,0 +1,195 @@
+
+    // Check that the balance is less or equal to the token balance
+    require(balance <= token.balanceOf(address(this)));
+
+    // Mark channel as closed and create closing request
+    closeTime = now.add(challengeTime);
+    closingBalance = balance;
+  }
+
+  /**
+   * @dev Close a channel with the agreement of the sender and receiver
+   * @param balance uint256, the final balance transfered of the channel
+   * @param balanceMsgSig bytes, the signature of the sender.
+    The msg signed by the receiver is generateBalanceHash(balance)
+   * @param closingSig bytes, the signature of the receiver
+    The msg signed by the sender is generateKeccak256(balanceMsgSig)
+   */
+  function cooperativeClose(uint256 balance, bytes balanceMsgSig, bytes closingSig) external {
+    // Derive receiver address from signature
+    require(receiver == keccak256(balanceMsgSig).recover(closingSig));
+
+    // Derive sender address from signed balance proof
+    address sender = getSignerOfBalanceHash(balance, balanceMsgSig);
+    require(sender == owner);
+
+    close(balance);
+  }
+
+  /**
+   * @dev Close a channel with an uncooperativeClose already requested
+   */
+  function closeChannel() external onlyOwner {
+    // Check that the closing request was created
+    require(closeTime > 0);
+
+    // Make sure the challengeTime has ended
+    require(block.timestamp > closeTime);
+
+    close(closingBalance);
+  }
+
+  /*
+   *  Public functions
+   */
+
+   /**
+    * @dev Get the channel info
+    */
+  function getInfo() public view returns (uint256, uint256, uint256) {
+    return (token.balanceOf(address(this)), closeTime, closingBalance);
+  }
+
+  /**
+   * @dev Get the signer of a balance hash signed
+   * @param balance uint256, the balance to hash
+   * @param msgSigned bytes, the balance hash signed
+   */
+  function getSignerOfBalanceHash(uint256 balance, bytes msgSigned) public view returns (address) {
+    bytes32 msgHash = generateBalanceHash(balance);
+    // Derive address from signature
+    address signer = msgHash.recover(msgSigned);
+    return signer;
+  }
+
+  /**
+   * @dev Generate a hash balance for an address
+   * @param balance uint256, the balance to hash
+   */
+  function generateBalanceHash(uint256 balance) public view returns (bytes32) {
+    return keccak256(receiver, balance, address(this));
+  }
+
+  /**
+   * @dev Generate a keccak256 hash
+   * @param message bytes, the mesage to hash
+   */
+  function generateKeccak256(bytes message) public pure returns(bytes32) {
+    return keccak256(message);
+  }
+
+  /*
+   *  Internal functions
+   */
+
+  /**
+   * @dev Close a channel
+   * @param finalBalance uint256, the final balance of the receiver
+   */
+  function close(uint256 finalBalance) internal {
+    uint256 tokenBalance = token.balanceOf(address(this));
+    require(finalBalance <= tokenBalance);
+
+    // Send balance to the receiver
+    require(token.transfer(receiver, finalBalance));
```

Comment:
Not sure if it's relevant, but if the token is an ERC223, the receiver could reenter at this point, presenting the same signed message from the sender, and extract twice the funds (or n times, if he reenters more than once). Or he could force the transfer to fail, making the closure of the channel impossible.

Would it be safer to change these `transfer` calls to `approvals`, so each user can later extract the balances themselves?

Reply:
On the other hand, given that both addresses need to sign messages with their private keys, they both are user addresses, and not contracts, so this sould not be a problem. Right?

Reply:
And also, I'm not sure if ERC223 is being much used at all

Reply:
Yes, I agree too, is not being used much, but the problem still there, and it can be solved changing transfer with approval. Regarding user signatures that should be a problem, we should add a check that verifies that the sender and receiver are no contracts in the constructor. This way we will make sure it will be used by user addresses.
WDYT @spalladino ?

Reply:
Agree to change the transfer into an approval, which could even use an `asyncSend` version for ERC20s that we don't yet have. As for user signatures, I think it's ok not to check them, since the channel cannot be used if that were the case (no signatures can be exchanged). Besides, a user could spawn a contract in that address after the channel was created, making the check pointless.

Reply:
done!